pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/relocation/RealtimeSegmentRelocationManager.java

```diff
@@ -0,0 +1,287 @@
+  }
+
+  /**
+   * Check all tables. Perform relocation of segments if table is realtime and relocation is required
+   */
+  public void runRelocation() {
+    if (!_pinotHelixResourceManager.isLeader()) {
+      LOGGER.info("Skipping realtime segment relocation, not leader!");
+      return;
+    }
+
+    LOGGER.info("Starting relocation of realtime segments");
+    List<String> allTableNames = _pinotHelixResourceManager.getAllTables();
+
+    for (String tableNameWithType : allTableNames) {
+
+      TableType tableType = TableNameBuilder.getTableTypeFromTableName(tableNameWithType);
+      if (tableType.equals(TableType.REALTIME)) {
+        LOGGER.info("Starting relocation of segments for table: {}", tableNameWithType);
+
+        TableConfig tableConfig = _pinotHelixResourceManager.getRealtimeTableConfig(tableNameWithType);
+
+        RealtimeTagConfig realtimeTagConfig = new RealtimeTagConfig(tableConfig, _helixManager);
+        if (!realtimeTagConfig.isRelocateCompletedSegments()) {
+          LOGGER.info("Skipping relocation of segments for {}", tableNameWithType);
+          return;
+        }
+
+        IdealState idealState = HelixHelper.getTableIdealState(_helixManager, tableNameWithType);
+        if (!idealState.isEnabled()) {
+          LOGGER.info("Skipping relocation of segments for {} since ideal state is disabled", tableNameWithType);
+          return;
+        }
+
+        Map<String, Map<String, String>> segmentNameToInstancesMap = relocateSegments(realtimeTagConfig, idealState);
+        if (!segmentNameToInstancesMap.isEmpty()) {
+          updateIdealState(tableNameWithType, segmentNameToInstancesMap);
+        }
+      }
+    }
+    LOGGER.info("Realtime segment relocation completed");
+  }
+
+  /**
+   * Given a realtime tag config and an ideal state, relocate the segments
+   * which are completed but still hanging around on consuming servers, one replica at a time
+   * @param  realtimeTagConfig
+   * @param idealState
+   */
+  protected Map<String, Map<String, String>> relocateSegments(RealtimeTagConfig realtimeTagConfig,
+      IdealState idealState) {
+    Map<String, Map<String, String>> segmentNameToInstancesMap;
+
+    List<String> consumingServers = _helixAdmin.getInstancesInClusterWithTag(_helixManager.getClusterName(),
+        realtimeTagConfig.getConsumingServerTag());
+    if (consumingServers.isEmpty()) {
+      throw new IllegalStateException(
+          "Found no realtime consuming servers with tag " + realtimeTagConfig.getConsumingServerTag());
+    }
+    List<String> completedServers = _helixAdmin.getInstancesInClusterWithTag(_helixManager.getClusterName(),
+        realtimeTagConfig.getCompletedServerTag());
+    if (completedServers.isEmpty()) {
+      throw new IllegalStateException(
+          "Found no realtime completed servers with tag " + realtimeTagConfig.getCompletedServerTag());
+    }
+
+    Map<String, Map<String, String>> mapFields = idealState.getRecord().getMapFields();
```

Comment:
Please add a TODO here to use the SegmentAssigmentStrategy to place newly completed segments